ansible_parent_role_name/path variables for when a role is being included by another role

Xaroth opened the pull request:
##### SUMMARY
Currently it is not possible for roles to detect when they are being directly included by another role.

This PR does the following:
- When a role is being included, set the `ansible_parent_role_name` to the name of the role that included this role.
- When a role is being included, set the `ansible_parent_role_path` to the path of the role that included this role.
- Document the `role_name` magic variable. It existed, but there was no existing documentation. Its companion `role_path` was already documented.

The main rationale for this is that we currently have a large set of roles that, in addition of providing with a state that is being managed, also expose a set of extra callable task files to allow other roles to easily perform tasks related to that role, without having to duplicate logic several dozen times.

During these includes, it is often required to send an identifier along, so that comments can indicate which role/feature requested these actions. Currently this is done by passing variables to each include_role action, but this proves to be prone to error.
The `role_name` variable already exists, but since this only stores the current executing role, it doesn't help trace down the source if the role was included otherwise.

The proposed new magic variables do specifically that; letting the included roles know who included them, so that it's easier to build conditions that only run (or should never run) when tasks are being included, and to be able to better debug the source of configuration.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
Magic Variables

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```
ansible 2.7.0
```
but this affects all versions.

##### ADDITIONAL INFORMATION

webknjaz commented:
Hi @Xaroth,

It looks like an architectural change, which I believe requires a broader discussion.
Could you please go ahead and create a proposal in https://github.com/ansible/proposals/issues/new pointing to this PR as a reference implementation?

Thanks in advance!

Xaroth commented:
@webknjaz Thanks, proposal sent in.

gundalow reviewed (commented):
Would be good to add some tests to defend this functionality

bcoca commented:
i would make this a list so you can trace back the full stack, not just the immediate parent

Xaroth commented:
@bcoca I updated the PR with your comments. I also added tests as per the request of @gundalow

Xaroth committed:
-Add: documentation for the ``ansible_parent_role_names`` magic variable
-Add: Test cases for ansible_parent_role_names and ansible_parent_role_paths
-Add: ansible_parent_role_names/paths variables for when a role is being included by another role.

Xaroth commented:
Rebased to fix merge conflicts, squashed into one commit for better readability.

jamescassell commented on docs/docsite/rst/reference_appendices/special_variables.rst:
```diff
@@ -25,6 +25,15 @@ ansible_inventory_sources
+ansible_parent_role_names
+    Similar to ``ansible_parent_role_name``, only contains a list of all parent roles, with the most recent role (i.e. the role that included this role) being the first item in the list.
+
+ansible_parent_role_path
```
we shoud add parent_role_paths to be consistent w/ parent_role_names

nitzmahone commented:
If we're listifying both path and name, maybe we should also ditch the non-list versions and just use those...

bcoca commented:
@nitzmahone that was my hope, but add this first, then deprecate the others

Xaroth commented:
Added the list of paths, as requested by @jamescassell .

mattclay commented on test/integration/targets/include_parent_role_vars/meta/main.yml:
```diff
@@ -0,0 +1 @@
+---
```
Why have an empty `meta/main.yml`?

mattclay commented on test/integration/targets/include_parent_role_vars/aliases:
```diff
@@ -0,0 +1,2 @@
+hidden/special_vars
+unsupported
```
Remove the `unsupported` alias as it's not needed.

mattclay commented on test/integration/targets/include_parent_role_vars/aliases:
```diff
@@ -0,0 +1,2 @@
+hidden/special_vars
```
This should just be `hidden`.

mattclay commented on test/integration/targets/special_vars/meta/main.yml:
```diff
@@ -1,2 +1,3 @@
 dependencies:
   - prepare_tests
+  - include_parent_role_vars
```
Is this needed for the test? It looks like it's just here to indicate the dependency is required. If that's the case, add an entry to the `aliases` file instead: `needs/target/include_parent_role_vars`

mattclay commented on test/integration/targets/special_vars/tasks/main.yml:
```diff
@@ -56,3 +56,8 @@
+- name: test ansible_parent_role_name(s) functionality by including another role
+  include_role:
+    name: include_parent_role_vars
+    tasks_from: included_by_other_role.yml
```
This test doesn't appear to have any relation to the `special_vars` integration test. It should probably be its own test target instead.

Xaroth replied:
The `ansible_parent_role_*` vars are special vars, as such, have a direct relation to being tested along with other special vars. However, since these special vars are only activated by including another role, I opted for including said other role from within the `special_vars` integration test.

mattclay commented on test/integration/targets/include_parent_role_vars/tasks/included_by_other_role.yml:
```diff
@@ -0,0 +1,32 @@
+
+- name: for display; our current parent tree
+  debug:
+    var: ansible_parent_role_names
```
It looks like this task can be removed without affecting the test results. Doing so will minimize test output.

mattclay commented on test/integration/targets/include_parent_role_vars/tasks/included_by_ourselves.yml:
```diff
@@ -0,0 +1,29 @@
+
+- name: for display; our current parent tree
+  debug:
+    var: ansible_parent_role_names
```
It looks like this task can be removed without affecting the test results. Doing so will minimize test output.

mattclay commented on test/integration/targets/include_parent_role_vars/tasks/included_by_ourselves.yml:
```diff
@@ -0,0 +1,29 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
```
If you're going to include a copyright and license notice for test files, you can use a shorter version instead:

```yaml
# Copyright 2019 Ansible Project
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

Xaroth replied:
I opted for using the same license notice as used in the parent test; `special_vars`, but this shorter one makes sense.

acozine commented on docs/docsite/rst/reference_appendices/special_variables.rst:
```diff
@@ -28,6 +28,18 @@ ansible_limit
+    When the current role is being executed by means of an include_role action, this variable contains the role name that included this role. Otherwise, this variable is not defined.
+
+ansible_parent_role_names
+    Similar to ``ansible_parent_role_name``, only contains a list of all parent roles, with the most recent role (i.e. the role that included this role) being the first item in the list.
```
I'm a bit confused by this. When a role has multiple parent roles, is the relationship a tree (Role A includes Role B, Role B includes Role C . . . Role C's parents are B first and then A)? Or will this show if one role is (mistakenly?) included more than once by different roles (Role A includes Role M, Role G also includes Role M . . . Role M's parents are G and A in order of the playbook)? This is more for my information than a comment on the PR or the docs.

Xaroth replied:
As mentioned on IRC; I expanded these sections a bit to explain the order of the ``ansible_parent_role_names`` and ``ansible_parent_role_paths`` variables. Since there is no area in the documentation about how to deal with _being included_ by other roles, there isn't really a spot for this PR to attach itself to. It would likely make more sense that a separate chapter be made specifically about creating reusable roles and how to deal with being included by other roles.

acozine commented:
I'd like to see a paragraph on this functionality added to the roles documentation, with a scenario that demonstrates how to use the new variables.

acozine reviewed (commented):
Oops, I left a comment rather than starting a review - I'll repeat it here. I'd like to see a paragraph with a scenario that demonstrates how to use these variables added to the roles documentation (https://docs.ansible.com/ansible/devel/user_guide/playbooks_reuse_roles.html).

sivel commented:
> @nitzmahone that was my hope, but add this first, then deprecate the others

@bcoca I'm confused.  Why add them here now, to then deprecate them. 

I believe what was said is that we should only accept the listified values, and not accept the singular also.

bcoca commented:
@sivel, add the new ones, once working and all seems good we deprecate the old ones as they are just a subset

sivel commented:
It was discussed in the core IRC meeting today, that this PR should only add the plural (list) variables.

The singular forms should be removed.

Xaroth commented:
As per the core meeting (sorry that I missed it, I was otherwise occupied):

- Removed the singular values ( `ansible_parent_role_name|path`).
- Removed tests including those singular values (the list var is still being tested).
- Removed documentation for the singular values.

mattclay commented on test/integration/targets/include_parent_role_vars/aliases:
```diff
@@ -0,0 +1 @@
+hidden
```
Include a comment indicating what this target is used for since this isn't something we normally do.

Xaroth replied:
Comment added, thanks.

jamescassell commented on test/integration/targets/include_parent_role_vars/tasks/main.yml:
```diff
@@ -0,0 +1,2 @@
+# Copyright 2019 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Why add an empty file?  Does something fail without it?

Xaroth replied:
I was under the impression that at least one */main.yml would have to exist for it to be considered an actual role.

That said, I will adjust it as I had considered a use case that needs to be tested that is currently not yet tested.

Xaroth commented:
I added an extra test case to ensure that include_tasks and import_tasks does not affect the `ansible_parent_role_names` vars.

jamescassell reviewed (approved):
shipit

This will bring very useful functionality, and appears to be well tested and documented.  For bonus points, the code change itself is trivial.

samdoran commented on docs/docsite/rst/reference_appendices/special_variables.rst:
```diff
@@ -31,6 +31,16 @@ ansible_loop
     The name of the value provided to ``loop_control.loop_var``. Added in ``2.8``
 
+ansible_parent_role_names
+    When the current role is being executed by means of an include_role action, this variable contains a list of all parent roles, with the most recent role (i.e. the role that included this role) being the first item in the list.
```
This seems to work with `import_role` as well. Docs should reflect that.

Xaroth replied:
Adjusted documentation to match.

samdoran commented on docs/docsite/rst/reference_appendices/special_variables.rst:
```diff
@@ -31,6 +31,16 @@ ansible_loop
+    For example: When role **A** includes role **B**, inside role B, ``ansible_parent_role_names`` will equal to ``['A']``. If role **B** then includes role **C**, the list becomes ``['B', 'A']``.
+
+ansible_parent_role_paths
+    When the current role is being executed by means of an include_role action, this variable contains a list of all parent roles, with the most recent role (i.e. the role that included this role) being the first item in the list.
```
Same as above: this magic var is not limited to `include_role`.

Xaroth replied:
Adjusted documentation to match.

samdoran commented on test/integration/targets/include_parent_role_vars/tasks/included_by_ourselves.yml:
```diff
@@ -0,0 +1,9 @@
+    that:
+      - "ansible_parent_role_names|length == 2"
+      - "ansible_parent_role_names[0] == 'include_parent_role_vars'"  # Since we included ourselves, we're the top level
+      - "ansible_parent_role_names[1] == 'special_vars'"
```
We also need to test `ansible_parent_role_paths`.

Xaroth replied:
Tests for `ansible_parent_role_paths` added.

samdoran commented on test/integration/targets/include_parent_role_vars/tasks/included_by_other_role.yml:
```diff
@@ -0,0 +1,12 @@
+- name: ensure our parent role tree to contain only our direct parent item
+  assert:
+    that:
+      - "ansible_parent_role_names == ['special_vars']"
```
We also need to test `ansible_parent_role_paths`

Xaroth replied:
As discussed on IRC, currently only checking for length equality to ansible_parent_role_paths as actual value checking is made difficult due to temp dirs.

samdoran commented on test/integration/targets/include_parent_role_vars/tasks/included_by_other_role.yml:
```diff
@@ -0,0 +1,12 @@
+      - "ansible_parent_role_names == ['special_vars']"
+
+- name: attempt to include ourselves
+  include_role:
```
We need tests for `import_role` as well.

Xaroth replied:
Added.

samdoran commented on test/integration/targets/special_vars/tasks/main.yml:
```diff
@@ -56,3 +56,12 @@
       - "(ansible_play_role_names + ansible_dependent_role_names)|unique|sort|list == ansible_role_names|sort|list"
+
+- name: test ansible_parent_role_name(s) functionality by including another role
+  include_role:
```
We need tests for `import_role` as well.

Xaroth replied:
Done.

Xaroth commented:
Added changes requested by @samdoran:

- Make documentation reflect that both `include_role` as well as `import_role` cause this behavior.
- Add links to the `include_role` and `import_role` modules from that documentation, so that it's easier to find those module specifics
- Added tests for `ansible_parent_role_paths`. Currently only checking for lenght equality to `ansible_parent_role_names` as the tempdirs created by `ansible-test` make it difficult to fully check path values.
- Added tests to ensure that both `ansible_parent_role_names` and `ansible_parent_role_paths` are undefined prior to roles being included/imported, and after they have been included/imported.
- Added tests to ensure that the values reset to their original values after importing/including a role.

These changes add a lot of tests to the mix (around 50 new actions for the `special_vars` tests), but this shouldn't affect test runtime by much.

samdoran reviewed (approved):
I'll merge this for 2.9 after we create a branch for 2.8.
